### ISSUE TITLE
Add defaults for `State.ERROR`

### ISSUE THREAD

**Issue by DavePearce** [ConsenSys/evm-dafny#534]
At the moment, `State.ERROR` is defined as follows:

```Dafny
datatype State = EXECUTING(evm:T)
        | ERROR(error:Error,gas:nat,data: Array<u8>)
```

However, it would be better to define it thusly:

```dafny
datatype State = EXECUTING(evm:T)
        | ERROR(error:Error,gas:nat := 0,data: Array<u8> := [])
```

However, I encountered a strange internal failure with Dafny doing that.

**Comment by DavePearce** [comment]
Ok, it does work but there are some places where the bug arises.  For example:

```Dafny
function Call(world: WorldState.T, ctx: Context.T, substate: SubState.T, codeAddress: u160, value: u256, gas: nat, depth: nat) : State
    // Sender account must exist
    requires world.Exists(ctx.sender) 

   ...

                    match Precompiled.Call(codeAddress,ctx.callData)
                    case None => ERROR(INVALID_PRECONDITION,0,[])
```

So, would be good to raise an issue and update these at some point.

**Comment by DavePearce** [comment]
This issue seems to be fixed in Dafny 4.0, and will be done as part of #531
